leaf-server/src/main/java/org/dreeam/leaf/util/list/OptimizedNonNullListArrayList.java

```diff
@@ -0,0 +1,428 @@
+
+        final E old = a[index];
+        final int numMoved = size - index - 1;
+
+        if (numMoved > 0) {
+            System.arraycopy(a, index + 1, a, index, numMoved);
+        }
+
+        a[--size] = null; // Clear reference for GC
+        invalidateHash();
+        return old;
+    }
+
+    @Override
+    public E set(final int index, final E k) {
+        Objects.requireNonNull(k, "Cannot set null element in this list");
+
+        if ((index | (size - 1 - index)) < 0) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+
+        final E old = a[index];
+        a[index] = k;
+        invalidateHash();
+        return old;
+    }
+
+    @Override
+    public void clear() {
+        if (defaultValue != null && size > 0) {
+            Arrays.fill(a, 0, size, defaultValue);
+        }
+        size = 0;
+        invalidateHash();
+    }
+
+    /**
+     * fast fill with default value
+     */
+    public void fillWithDefault() {
+        if (hasFlag(HAS_DEFAULT_FLAG) && size > 0) {
+            Arrays.fill(a, 0, size, defaultValue);
+            invalidateHash();
+        }
+    }
+
+    // === SEARCH OPERATIONS ===
+
+    @Override
+    public int indexOf(final Object k) {
+        if (k == null) return -1;
+
+        final Object[] array = this.a;
+        final int length = this.size;
+
+        int i = 0;
+        for (; i < length - 3; i += 4) {
+            if (k.equals(array[i])) return i;
+            if (k.equals(array[i + 1])) return i + 1;
+            if (k.equals(array[i + 2])) return i + 2;
+            if (k.equals(array[i + 3])) return i + 3;
+        }
+
+        for (; i < length; i++) {
+            if (k.equals(array[i])) return i;
+        }
+
+        return -1;
+    }
+
+    @Override
+    public int lastIndexOf(final Object k) {
+        if (k == null) return -1;
+
+        final Object[] array = this.a;
+
+        for (int i = size - 1; i >= 3; i -= 4) {
+            if (k.equals(array[i])) return i;
+            if (k.equals(array[i - 1])) return i - 1;
+            if (k.equals(array[i - 2])) return i - 2;
+            if (k.equals(array[i - 3])) return i - 3;
+        }
+
+        for (int i = Math.min(size - 1, 3); i >= 0; i--) {
+            if (k.equals(array[i])) return i;
+        }
+
+        return -1;
+    }
+
+    @Override
+    public int hashCode() {
+        if (hashValid) return cachedHashCode;
+
+        int h = 1;
+        final Object[] array = this.a;
+        final int length = this.size;
+
+        int i = 0;
+        for (; i < length - 3; i += 4) {
+            h = 31 * h + array[i].hashCode();
+            h = 31 * h + array[i + 1].hashCode();
+            h = 31 * h + array[i + 2].hashCode();
+            h = 31 * h + array[i + 3].hashCode();
+        }
+
+        for (; i < length; i++) {
+            h = 31 * h + array[i].hashCode();
+        }
+
+        cachedHashCode = h;
+        hashValid = true;
+        return h;
+    }
```

Comment:
```suggestion
    @Override
    public int hashCode() {
        if (hashValid) return cachedHashCode;
        
        final Object[] array = this.a;
        final int length = this.size;
        int h;
        int i = 0;
        if (length >=4) {

            int h1 = 1;
            int h2 = 3;
            int h3 = 7;
            int h4 = 11;

            for (; i < length - 3; i += 4) {
                h1 = 31 * h1 + array[i].hashCode();
                h2 = 31 * h2 + array[i + 1].hashCode();
                h3 = 31 * h3 + array[i + 2].hashCode();
                h4 = 31 * h4 + array[i + 3].hashCode();
            }
            h = h1 + h2 + h3 + h4;

        } else {
            h = 1;
        }

        for (; i < length; i++) {
            h = 31 * h + array[i].hashCode();
        }

        cachedHashCode = h;
        hashValid = true;
        return h;
    }
```

Make it even faster by making it so the hashcode additions can be pipelined in parallel.